### PR TITLE
Minesweeper Mine Detection

### PR DESCRIPTION
**Pull request by TheSilverHornet**
In which minelayers can see enemy mines around them, allowing their mine clearance attack to actually be game useful. No more are they only half a tool!

_Remember to give enemy team a resourcegen, so they have power to cloak their mines, otherwise -anything- detects them._

Screenshot (462)

### CODE REVIEW

File: luarules/gadgets/unit_minesweeper_detection.lua [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+
+if UnitDefNames.armmine1 then
+    mineIDs[UnitDefNames.armmine1.id] = true
+end
+if UnitDefNames.armmine2 then
+    mineIDs[UnitDefNames.armmine2.id] = true
+end
+if UnitDefNames.armmine3 then
+    mineIDs[UnitDefNames.armmine3.id] = true
```
Comment (sprunk) [inline luarules/gadgets/unit_minesweeper_detection.lua:89]: use customparams

File: luarules/gadgets/unit_minesweeper_detection.lua [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+local minesweeperIDs = {}
+local minesweepers = {}
+local mineIDs = {}
+
+--Spring.Echo('hornet ms loaded')
+
+local teamList = Spring.GetTeamList()
+local teamIDs = {}
+
+local revealedMines = {}=
```
Comment (WatchTheFort) [inline luarules/gadgets/unit_minesweeper_detection.lua:31]: Syntax error on this line
```suggestion
local revealedMines = {}
```

File: luarules/gadgets/unit_minesweeper_detection.lua (hunk 1 above)
Comment (TheSilverHornet) [reply luarules/gadgets/unit_minesweeper_detection.lua]: Good idea, done so

**Comment by WatchTheFort** [comment]
> _Note that you cannot spawn a minelayer to test, you have to spawn a veh lab, then build one._

What is the cause of this, and is there a way to fix it?

**Comment by TheSilverHornet** [comment]
it's only an issue when testing, not a concern in real games. possibly one of the meta 'created' hooks might work around it

**Comment by sprunk** [comment]
gadget uses UnitFinished instead of UnitCreated, and spawned units never need to be finished so they don't produce the callin

**Comment by TheSilverHornet** [comment]
@sprunk , would metaunitadded be better, hopefully that should catch all cases? I didn't know it existed when I wrote this, PR is a month old

**Comment by sprunk** [comment]
1) yes, then it will also handle cases where a minelayer changes team e.g. via capture
2) remember to handle nanoframes in that case (nanoframes of minelayers probably shouldn't detect things until they are ready)

**Comment by TheSilverHornet** [comment]
2 is unlikely to occur in real games, but if you have a concise solution I'd love to add it

**Comment by sprunk** [comment]
On second thought, I just checked and `/give` correctly produces a UnitFinished event. So idk why it would fail, but MetaUnitAdded is a good idea anyway due to capture/share.

**Comment by TheSilverHornet** [comment]
Updated to use meta instead

**Comment by Beherith** [comment]
1. Periodic minesweeper GetUnitsInCylinder should be: 
A. incongruent with frame rate. 
B. Staggered at least on a per-team basis
2. Use unitFinished and dont make nano sweepers reveal mines.
3. Is this modoption gated?
4. Has it passed through GDT? 
5. Note that website documentation or other flavor items might need updating. 
6. For empty checking of hash tables use next() instead of table.count

**Pull request merged**